### PR TITLE
[CARBONDATA-938] Prune partitions for filter query on partition column

### PR DESCRIPTION
**Pull request by QiangCai**
1. generate partition filter by basing on pushed-down filter

2. apply partition filter to get the map of required partition (use BitSet)
support EqualTo, In, GreaterThan/EqualTo(>, >=), LessThan/EqualTo(<, <=)

3. in CarbonInputFormat, use partition map to skip some non-required splits

4. add query test case

Prune Partitions as follow:
image

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2044/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2048/

**Comment by QiangCai** [comment]
@jackylk @ravipesala @lionelcao please review pr

File: core/src/main/java/org/apache/carbondata/core/scan/filter/partition/DefaultPartitionFilterImpl.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public class DefaultPartitionFilterImpl implements PartitionFilterIntf {
+
+  /**
+   * true: all partitions are required
+   * false: no partition is required
+   */
+  private boolean isContainAll;
+
+  public DefaultPartitionFilterImpl(boolean isContainAll) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/partition/DefaultPartitionFilterImpl.java]: There is only one instance creation whose `isContainAll` is `false`
Can you make another implementation for it? I think the name `isContainAll` is not intuitive.
```
      case FALSE:
        return new DefaultPartitionFilterImpl(false);
```

File: hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java [hunk 2]
```diff
@@ -255,11 +258,32 @@ private static AbsoluteTableIdentifier getAbsoluteTableIdentifier(Configuration
       if (null == carbonTable) {
         throw new IOException("Missing/Corrupt schema file for table.");
       }
+
       CarbonInputFormatUtil.processFilterExpression(filter, carbonTable);
+
+      // prune partitions for filter query on partition table
+      BitSet partitionMap = null;
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java]: suggest to rename all `partitionMap` to `matchedPartitions` in all functions

File: hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java [hunk 3]
```diff
@@ -255,11 +258,32 @@ private static AbsoluteTableIdentifier getAbsoluteTableIdentifier(Configuration
...
+        PartitionInfo partitionInfo = carbonTable.getPartitionInfo(carbonTable.getFactTableName());
+        if (null != partitionInfo) {
+          Partitioner partitioner = PartitionUtil.getPartitioner(partitionInfo);
+          partitionMap = new FilterExpressionProcessor()
+              .getFilteredPartitions(filter, partitionInfo, partitioner);
+          if (partitionMap.cardinality() == 0) {
+            // no partition is required
+            return new ArrayList<InputSplit>();
+          }
+          if (partitionMap.cardinality() == partitioner.numPartitions()) {
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java]: I feel it is more readable to use BitSet with all `1` than `null`

Comment (QiangCai) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java]: If all partitions are matched or not matched, we can release BitSet object.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/FilterExpressionProcessor.java [hunk 4]
```diff
@@ -137,6 +155,119 @@ public FilterResolverIntf getFilterResolver(Expression expressionTree,
...
+    return createPartitionFilterTree(expressionTree, partitionInfo).applyFilter(partitioner);
+  }
+
+  /**
+   * create partition filter by basing on pushed-down filter
+   * @param expressionTree
+   * @param partitionInfo
+   * @return
+   */
+  public PartitionFilterIntf createPartitionFilterTree(Expression expressionTree,
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/FilterExpressionProcessor.java]: ok

File: core/src/main/java/org/apache/carbondata/core/scan/filter/partition/DefaultPartitionFilterImpl.java (hunk 1 above)
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/partition/DefaultPartitionFilterImpl.java]: We need to create BitSet for each filter. We will invoke logical operate on these BitSets.  It means filters can't share the same BitSet.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2091/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2094/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2095/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**